File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int rows = matrix.size();
        int cols = matrix[0].size();
        
        // binary search - row
        int top = 0;
        int bot = rows - 1;
        while (top <= bot) {
            int mid = top + (bot - top) / 2;
            if (matrix[mid][cols - 1] < target) {
                top = mid + 1;
            } else if (matrix[mid][0] > target) {
                bot = mid - 1;
            } else {
                break;
            }
        }
        
        if (!(top <= bot)) return false;  // none of the rows contain the target
        int row = (top + bot) / 2;  // the target row
        
        // binary search - column
        int left = 0;
        int right = cols - 1;
        
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (matrix[row][mid] < target) {
                left = mid + 1;
            } else if (matrix[row][mid] > target) {
                right = mid - 1;
            } else {
                return true;  // target exists
            }
        }
        return false;
    }
};

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        rows, cols = len(matrix), len(matrix[0])
        
        # binary search - row
        top, bot = 0, rows - 1
        while top <= bot:
            row = top + (bot - top) // 2
            if matrix[row][-1] < target:
                top = row + 1
            elif matrix[row][0] > target:
                bot = row - 1
            else:
                break
        
        if not(top <= bot): # none of the rows contain the target value
            return False
        row = (top + bot) // 2
            
        # binary search - column
        left, right = 0, cols - 1
        while left <= right:
            mid = left + (right - left) // 2
            if matrix[row][mid] < target:
                left = mid + 1
            elif matrix[row][mid] > target:
                right = mid - 1
            else:
                return True
        return False